src/NumericalAlgorithms/Interpolation/ZeroCrossingPredictor.hpp

```diff
@@ -0,0 +1,78 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <deque>
+
+#include "DataStructures/DataVector.hpp"
+
+/// \cond
+namespace PUP {
+class er;
+}  // namespace PUP
+/// \endcond
+
+namespace intrp {
+
+/// \ingroup NumericalAlgorithmsGroup
+/// \brief A class that predicts when a function crosses zero
+class ZeroCrossingPredictor {
+ public:
+  /// Uses at most max_size times for the fit; throws away old
+  /// times as new times are added.
+  /// The is_valid function returns false until min_size times
+  /// have been added.  min_size must be at least 3.
+  ZeroCrossingPredictor(size_t min_size, size_t max_size);
+
+  ZeroCrossingPredictor() = default;
+  ZeroCrossingPredictor(const ZeroCrossingPredictor& /*rhs*/) = delete;
+  ZeroCrossingPredictor& operator=(const ZeroCrossingPredictor& /*rhs*/) =
+      delete;
+  ZeroCrossingPredictor(ZeroCrossingPredictor&& /*rhs*/) = default;
+  ZeroCrossingPredictor& operator=(ZeroCrossingPredictor&& rhs) = default;
+  ~ZeroCrossingPredictor() = default;
+
+  /// Adds a data point at time t to the ZeroCrossingPredictor.
+  void add(double t, const DataVector& data_at_time_t);
+  void add(double t, DataVector&& data_at_time_t);
```

Comment:
Is there a reason for these two overloads?  Based on the implementations, I think having a single by-value function would actually be appropriate.

Reply:
I added the 2nd overload to avoid copying a DataVector in the case where we compute that DataVector only for the purpose of passing it into ZeroCrossingPredictor::add.      Changing to a single by-value function would mean one extra copy (only for that case).  Should I make the change?

Reply:
I'm confused.  Wouldn't a by-value implementation do zero copies if the caller moved its object into the argument?

Reply:
Oh, yes you are correct.  Sorry I was confused.